### PR TITLE
Minor: deprecate ScalarValue::get_datatype()

### LINKED ISSUE #7506: Deprecate `ScalarValue::get_datatype()`
### Is your feature request related to a problem or challenge?

We introduced `ScalarValue::data_type()` in https://github.com/apache/arrow-datafusion/pull/7492 to be consistent with other APIs for getting `DataType`. 

However the previous `ScalarValue::get_datatype()`

https://github.com/apache/arrow-datafusion/pull/7492#event-10314345235

### Describe the solution you'd like

1. Mark `ScalarValue::get_datatype()` deprecated (like this): 

https://github.com/apache/arrow-datafusion/blob/93c209f1b5d0a17b2aa5c6743cbd3cb189a406c8/datafusion/expr/src/table_source.rs#L81-L87

2. Update the uses in the datafusion codebase to pass clippy (update to use `ScalarValue::data_type()`)

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Weijun-H**
## Which issue does this PR close?

Closes #7506

### CODE REVIEW

**Comment by alamb** [comment]
Thanks @Weijun-H 🙏

**Comment by alamb** [comment]
Looks like there are a few more places to fix up

**Comment by Weijun-H** [comment]
Hello @alamb, all the places have been fixed.

**Review by yjshen — APPROVED** [review]

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -1152,6 +1152,7 @@ impl ScalarValue {
     /// Getter for the `DataType` of the value.
     ///
     /// Suggest using  [`Self::data_type`] as a more standard API
+    #[deprecated(since = "30.0.0", note = "use data_type instead")]
```
Comment (viirya) [inline datafusion/common/src/scalar.rs:1155]:
```suggestion
    #[deprecated(since = "31.0.0", note = "use data_type instead")]
```

**Review by viirya — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Pull request merged**